modules/Bio/EnsEMBL/Hive/HivePipeline.pm

```diff
@@ -544,151 +557,197 @@ sub print_diagram {
+                my $tweakStructure;
+                $tweakStructure->{Object}->{Type} = "Analysis";
                 my $analysis_name = $analysis->logic_name;
-
+                $tweakStructure->{Object}->{Id} = $analysis->dbID + 0;
+                $tweakStructure->{Object}->{Name} = $analysis->logic_name;
                 if( $attrib_name eq 'wait_for' ) {
-
+                    $tweakStructure->{Return}->{Field} = 'wait_for';
                     my $cr_collection   = $self->collection_of( 'AnalysisCtrlRule' );
                     my $acr_collection  = $analysis->control_rules_collection;
-
+                    $tweakStructure->{Return}->{OldValue} = join(', ', map { $_->condition_analysis_url } @$acr_collection );
                     if($operator eq '?') {
-                        print "Tweak.Show    \tanalysis[$analysis_name].wait_for ::\t[".join(', ', map { $_->condition_analysis_url } @$acr_collection )."]\n";
+                        $tweakStructure->{Action} = "SHOW";
+                        $tweakStructure->{Return}->{NewValue} = $tweakStructure->{Return}->{OldValue};
+                        push @response, "Tweak.Show  \tanalysis[$analysis_name].wait_for ::\t[".join(', ', map { $_->condition_analysis_url } @$acr_collection )."]\n";
                     }
 
                     if($operator eq '#' or $operator eq '=') {     # delete the existing rules
+                        $tweakStructure->{Action} = "DELETE";
+                        $tweakStructure->{Return}->{NewValue} = undef;
                         foreach my $c_rule ( @$acr_collection ) {
                             $cr_collection->forget_and_mark_for_deletion( $c_rule );
                             $need_write = 1;
 
-                            print "Tweak.Deleting\t".$c_rule->toString." --> (missing value)\n";
+                            push @response, "Tweak.Deleting\t".$c_rule->toString." --> (missing value)\n";
                         }
                     }
 
                     if($operator eq '=' or $operator eq '+=') {     # create new rules
+                        $tweakStructure->{Action} = "SET";
+                        $tweakStructure->{Return}->{NewValue} = $new_value;
                         Bio::EnsEMBL::Hive::Utils::PCL::parse_wait_for($self, $analysis, $new_value);
                         $need_write = 1;
                     }
 
                 } elsif( $attrib_name eq 'flow_into' ) {
-
+                    $tweakStructure->{Return}->{Field} = 'flow_into';
+                    $tweakStructure->{Return}->{OldValue} = $analysis->logic_name;
                     if($operator eq '?') {
                         # FIXME: should not recurse
-                        $analysis->print_diagram_node($self, '', {});
+                        $tweakStructure->{Action} = 'SHOW';
+                        $tweakStructure->{Return}->{NewValue} = $tweakStructure->{Return}->{OldValue};
+                        #$analysis->print_diagram_node($self, '', {});
```

Comment:
`print_diagram_node` is not great because it recurses, but that's the only thing we have.

* Either we keep it, but (1) the FIXME comment must be moved closer and (2) the call must be wrapped something that captures STDOUT in order to add it to `@response`
* Or we admit we cannot print dataflow rules even in text mode, and we push a warning / explanation to `@response`

Reply:
There is very poor situation with output at all, in this particular file we have 4 type for output:
1. Direct print
2. Push to warnings with print at the end
3. Push to response that is returned
4. Making a hash, returned also (for further JSON)

That's why I have created (and it's pull request) formatter. 
in Formatter it can be added JSON, error, warning, info or custom output function (as here print_diagram_node)  - and printed in needed format in necessary time.

So in scope of this branch - we resolve problem - to get all print out of HivePipeline, that's why we can't keep here print_diagram_node.

In order to refactor it - we will apply formatter here eventually, and it will allow to keep print_diagram_node